Shape_detection/include/CGAL/Shape_detection/Region_growing/Region_growing_on_polygon_mesh/Polyline_graph.h

```diff
@@ -0,0 +1,363 @@
+    being a `PolygonMesh` into planar regions.
+
+    \tparam GeomTraits
+    a model of `Kernel`
+
+    \tparam PolygonMesh
+    a model of `FaceListGraph`
+
+    \tparam FaceRange
+    a model of `ConstRange` whose iterator type is `RandomAccessIterator` and
+    value type is the face type of a polygon mesh
+
+    \tparam EdgeRange
+    a model of `ConstRange` whose iterator type is `RandomAccessIterator` and
+    value type is the edge type of a polygon mesh
+
+    \tparam VertexToPointMap
+    a model of `LValuePropertyMap` whose key type is the vertex type of a polygon mesh and
+    value type is `Kernel::Point_3`
+
+    \cgalModels `NeighborQuery`
+  */
+  template<
+  typename GeomTraits,
+  typename PolygonMesh,
+  typename FaceRange = typename PolygonMesh::Face_range,
+  typename EdgeRange = typename PolygonMesh::Edge_range,
+  typename VertexToPointMap = typename boost::property_map<PolygonMesh, CGAL::vertex_point_t>::type>
+  class Polyline_graph {
+
+  private:
+    struct PEdge {
+      std::size_t index = std::size_t(-1);
+      std::set<std::size_t> sneighbors;
+      std::set<std::size_t> tneighbors;
+      std::pair<int, int> regions;
+    };
+
+  public:
+    /// \name Types
+    /// @{
+
+    /// \cond SKIP_IN_MANUAL
+    using Traits = GeomTraits;
+    using Face_graph = PolygonMesh;
+    using Face_range = FaceRange;
+    using Edge_range = EdgeRange;
+    using Vertex_to_point_map = VertexToPointMap;
+    /// \endcond
+
+    /*!
+      a model of `ConstRange` whose iterator type is `RandomAccessIterator` and
+      value type is the internal edge type.
+    */
+    using Segment_range = std::vector<PEdge>;
+
+    /*!
+      a model of `ReadablePropertyMap` whose key type is the value type of `Segment_range`
+      and value type is `Kernel::Segment_3`.
+    */
+    using Segment_map = internal::Polyline_graph_segment_map<PEdge, Face_graph, Edge_range, Vertex_to_point_map>;
+
+    /// @}
+
+  private:
+    using Face_to_region_map = internal::Item_to_region_index_map;
+    using Face_to_index_map = internal::Item_to_index_property_map<Face_range>;
+    using Edge_to_index_map = internal::Item_to_index_property_map<Edge_range>;
+
+  public:
+    /// \name Initialization
+    /// @{
+
+    /*!
+      \brief initializes all internal data structures.
+
+      \param pmesh
+      an instance of a `PolygonMesh` that represents a polygon mesh
+
+      \param regions
+      a set of regions, where each region is an `std::vector<std::size_t>` and
+      contains indices of all faces in a polygon mesh been identified as the same plane.
+
+      \param vertex_to_point_map
+      an instance of `VertexToPointMap` that maps a polygon mesh
+      vertex to `Kernel::Point_3`
+
+      \pre `faces(pmesh).size() > 0`
+      \pre `edges(pmesh).size() > 0`
+      \pre `regions.size() > 0`
+    */
+    Polyline_graph(
+      const PolygonMesh& pmesh,
+      const std::vector< std::vector<std::size_t> >& regions,
```

Comment:
this should be a face -> id property map?

Reply:
Exactly. This is what I shortly mentioned in the description of this PR. In fact, I will change the API of this class because it is not good now. I have realized it already when I submitted the PR and worked on the docs.